slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/DbSlashingProtection.java

```diff
@@ -43,45 +68,126 @@ public boolean maySignAttestation(
+    if (surroundingAttestation.isPresent()) {
+      LOG.warn(
+          "Detected surrounding attestation {} for attestation signingRoot={} sourceEpoch={} targetEpoch={} publicKey={}",
+          surroundingAttestation.get(),
+          signingRoot,
+          sourceEpoch,
+          targetEpoch,
+          publicKey);
       return false;
     }
 
+    // check that no previous vote is surrounded by attestation
+    final Optional<SignedAttestation> surroundedAttestation =
+        signedAttestationsDao.findSurroundedAttestation(handle, id, sourceEpoch, targetEpoch);
+    if (surroundedAttestation.isPresent()) {
+      LOG.warn(
+          "Detected surrounded attestation {} for attestation signingRoot={} sourceEpoch={} targetEpoch={} publicKey={}",
+          surroundedAttestation.get(),
+          signingRoot,
+          sourceEpoch,
+          targetEpoch,
+          publicKey);
+      return false;
+    } else {
+      return true;
+    }
+  }
+
+  @Override
+  public boolean maySignBlock(
+      final Bytes publicKey, final Bytes signingRoot, final UInt64 blockSlot) {
+    final long validatorId = validatorId(publicKey);
     return jdbi.inTransaction(
         h -> {
-          final long id = validatorId.get();
           final Optional<SignedBlock> existingBlock =
-              signedBlocksDao.findExistingBlock(h, id, blockSlot);
+              signedBlocksDao.findExistingBlock(h, validatorId, blockSlot);
+
           // same slot and signing_root is allowed for broadcasting previously signed block
           // otherwise if slot and different signing_root then this is a double block proposal
-          final boolean maySign =
-              existingBlock.isEmpty() || existingBlock.get().getSigningRoot().equals(signingRoot);
-          if (maySign) {
-            final SignedBlock signedBlock = new SignedBlock(id, blockSlot, signingRoot);
+          if (existingBlock.isEmpty() || existingBlock.get().getSigningRoot().equals(signingRoot)) {
+            final SignedBlock signedBlock = new SignedBlock(validatorId, blockSlot, signingRoot);
             signedBlocksDao.insertBlockProposal(h, signedBlock);
+            return true;
+          } else {
+            LOG.warn("Detected double signed block {} for {}", existingBlock.get(), publicKey);
+            return false;
           }
-          return maySign;
         });
   }
 
   @Override
   public void registerValidators(final List<Bytes> validators) {
     jdbi.useTransaction(
         h -> {
-          final List<Validator> registeredValidators =
+          final List<Validator> existingRegisteredValidators =
               validatorsDao.retrieveValidators(h, validators);
           final List<Bytes> validatorsMissingFromDb = new ArrayList<>(validators);
-          registeredValidators.forEach(v -> validatorsMissingFromDb.remove(v.getPublicKey()));
-          validatorsDao.registerValidators(h, validatorsMissingFromDb);
+          existingRegisteredValidators.forEach(
```

Comment:
nit: its personal - but I prefer to read validatorsMissingFromDb.removeAll(existingRegisteredValidators.stream.map(Validator::publicKey)) ...(ymmv)

Reply:
done